coffee/rewriter.py

```diff
@@ -411,593 +446,81 @@ def preevaluate(self):
-        def process(symbols, should_factorize, op=None):
-            operands = set(s for s in symbols if should_factorize(s))
-            factors = set(s for s in symbols if not should_factorize(s))
-            return ExpressionFactorizer.Term(operands, factors, op)
-
-    def __init__(self, stmt):
-        self.stmt = stmt
+            On Optimality of Finite Element Integration, Luporini et. al.
+        """
+        lda = ldanalysis(self.expr_info.linear_loops[0], key='symbol', value='dim')
+        sg_visitor = SharingGraph(self.expr_info, lda)
+
+        # Maximize the visibility of linear symbols
+        sgraph, mapper = sg_visitor.visit(self.stmt.rvalue)
+        if 'topsum' in mapper:
+            self.expand(mode='linear', subexprs=[mapper['topsum']], not_aggregate=True)
+            sgraph, mapper = sg_visitor.visit(self.stmt.rvalue)
+
+        nodes, edges = sgraph.nodes(), sgraph.edges()
+
+        if self.expr_info.is_linear:
+            self.factorize(mode='adhoc', adhoc={n: [] for n in nodes})
+            self.licm('only_outlinear')
+        elif self.expr_info.is_bilinear:
+            # Resort to an ILP formulation to find out the best factorization candidates
+            if not (nodes and all(sgraph.degree(n) > 0 for n in nodes)):
+                self.factorize(mode='heuristic')
+                self.licm(mode='only_outlinear')
+                return
+            # Note: need to use short variable names otherwise Pulp might complain
+            nodes_vars = {i: n for i, n in enumerate(nodes)}
+            vars_nodes = {n: i for i, n in nodes_vars.items()}
+            edges = [(vars_nodes[i], vars_nodes[j]) for i, j in edges]
+
+            # ... declare variables
+            x = ilp.LpVariable.dicts('x', nodes_vars.keys(), 0, 1, ilp.LpBinary)
+            y = ilp.LpVariable.dicts('y', [(i, j) for i, j in edges] + [(j, i) for i, j in edges],
+                                     0, 1, ilp.LpBinary)
+            limits = defaultdict(int)
+            for i, j in edges:
+                limits[i] += 1
+                limits[j] += 1
+
+            # ... define the problem
+            prob = ilp.LpProblem("Factorizer", ilp.LpMinimize)
+
+            # ... define the constraints
+            for i in nodes_vars:
+                prob += ilp.lpSum(y[(i, j)] for j in nodes_vars if (i, j) in y) <= limits[i]*x[i]
+
+            for i, j in edges:
+                prob += y[(i, j)] + y[(j, i)] == 1
+
+            # ... define the objective function (min number of factorizations)
+            prob += ilp.lpSum(x[i] for i in nodes_vars)
+
+            # ... solve the problem
+            prob.solve(ilp.GLPK(msg=0))
+
+            # Finally, factorize and hoist (note: the order in which factorizations are carried
+            # out is crucial)
+            nodes = [nodes_vars[n] for n, v in x.items() if v.value() == 1]
+            other_nodes = [nodes_vars[n] for n, v in x.items() if nodes_vars[n] not in nodes]
```

Comment:
If I read the docs for pulp correctly, you're meant to pull the answer and the objective out of the solved problem.  Does it also update variables "in place"?